### DISCUSSION TITLE
Copy Mesh: `Error: Cannot connect disconnected graphs.`

### DISCUSSION THREAD

**Discussion by marwie** [donmccurdy/glTF-Transform#1299] [General]
I'm currently experimenting with creating a new GLB on the fly and copying some of the meshes to the new GLB for processing.

Currently having this code which however results in `Cannot connect disconnected graphs`

Any hints on why this happens/can be fixed would be greatly appreciated :)

```ts
const newDoc = new Document();
	newDoc.createScene();
	newDoc.createBuffer();

	const node = newDoc.createNode();

	const newMesh = newDoc.createMesh();
	node.setMesh(newMesh);

	newMesh.setName(mesh.getName());
	newMesh.copy(mesh);
```

**Comment by donmccurdy** [comment]
I think the issue is with this line:

```javascript
newMesh.copy(mesh);
```

... I assume `mesh` is part of a different Document than `newMesh`? The implementation of .copy() involves having the destination mesh assign itself references to the primitives of the source mesh, which refer to accessors, which refer to buffers ... and these resources can't be shared by two Documents.

Something like the proposal in https://github.com/donmccurdy/glTF-Transform/issues/924 might get you closer.

**Reply by marwie** [reply]
Thanks for the very fast reply. Yes they are - getting closer by copying primitive/attributes now but if it works with a simple copy and is handled properly under the hood I'd very much prefer that :) gonna try it

**Reply by marwie** [reply]
Ok `newMesh.copy(mesh, (p) => p.clone());` still results in the same error unfortunately

**Reply by donmccurdy** [reply]
I think the workaround i suggested in the post above (before editing it out ...) was incorrect sorry! The [implementation of `document.clone()`](https://github.com/donmccurdy/glTF-Transform/blob/main/packages/core/src/document.ts#L141-L142) is not a one-liner, but extracting bits of that should work, and it basically involves copying or reconstructing the primitives and attributes first, then doing a copy() on the mesh with a callback that returns the equivalent primitives from the new document. Not necessarily simpler than manually rebuilding the mesh at that point, though.

**Reply by marwie** [reply]
Thanks for the pointer. I got a lot closer (currently manually copying the mesh/primitive/accessors) - just got to figure out what `"message": "Entity cannot be empty."` in my output GLB is referring to

**Reply by marwie** [reply]
Ah - i just missed adding the node to the scene :)

**Reply by marwie** [reply]
![Capto_Capture 2024-03-07_05-25-34_AM](https://github.com/donmccurdy/glTF-Transform/assets/5083203/fbab4856-b0cd-4477-a814-a69c5c866421)

happy with the result for a quick hack - thanks again :)
